lib/tom_queue/worker_supervisor.rb

```diff
@@ -0,0 +1,84 @@
+module TomQueue
+  class WorkerSupervisor
+
+    attr_accessor :processes, :before_fork, :after_fork
+
+    def initialize
+      @processes = {}
+      @before_fork = -> { }
+      @after_fork = -> { }
+    end
+
+    def supervise(as:, count: 1, &block)
+      count.times do |i|
+        processes["#{as}#{i + 1}"] = block
+      end
+    end
+
+    def run
+      @stop = false
+      process_ids = {}
+      before_fork.call
+      loop do
+        processes_to_start = (processes.keys - process_ids.values)
+
+        processes_to_start.each do |name|
+          $stdout.puts "Starting '#{name}'..."
+          process_id = fork_process(name)
+          setup_signal_handling_for_process(process_id)
```

Comment:
Could we move the logging/signal handling stuff inside the `fork_process` method? Just means this method is nice and easy to read.

Reply:
Ah - no you'll want two signal handler setups - one in the parent process before you enter the loop, and one in the child after the fork that clears and resets the default signal handler.